### PR TITLE
Replaced the accessors that return optional string_view objects with two separate functions.

### PR DESCRIPTION
**Pull request by glynos**
This is what we talked about earlier (or last night, if you are reading about this when you wake up).

I removed the use of optional, and updated all the tests. For me, all tests are green. So the following is up for discussion:

1. Do we need "bool has_x() const" if string_view::data() can be tested against nullptr;
2. What do we do with "template <typename intT> intT port() const"?

For 1, I think (weakly) that the answer is "yes"
For 2, I think throw an exception on error

I'm already comfortable with the idea of no longer using optional in this context, my experience already is that is much easier for the user to use and understand.

### CODE REVIEW

**Comment by glynos** [comment]
P.S. If we agree, I want to start integrating this (as a submodule) into the cpp-netlib

**Comment by deanberris** [comment]
> Do we need "bool has_x() const" if string_view::data() can be tested against nullptr;

I think so. It's useful to have these specific checks.

> What do we do with "template intT port() const"?

I think we can turn this into a `string_view` too. If users want to turn this into a number, they can choose the many ways to do it.

> I'm already comfortable with the idea of no longer using optional in this context, my experience already is that is much easier for the user to use and understand.

SGTM

File: src/network/uri/uri.hpp [hunk 1]
```diff
@@ -215,68 +215,125 @@ class uri {
   const_iterator end() const;
 
   /**
+   * \brief Tests whether this URI has a scheme component.
+   * \return \c true if the URI has a scheme, \c false otherwise.
+   */
+  bool has_scheme() const;
```
Comment (deanberris) [inline src/network/uri/uri.hpp]: These probably should be `noexcept` alongside const.

File: src/network/uri/uri.hpp [hunk 2]
```diff
@@ -215,68 +215,125 @@ class uri {
...
+
+  /**
    * \brief Returns the URI scheme.
-   * \return The scheme, if it exists, or nullopt.
+   * \return The scheme.
+   * \pre has_scheme()
+   */
+  string_view scheme() const;
```
Comment (deanberris) [inline src/network/uri/uri.hpp]: The same (`noexcept`) for these accessors.

File: src/network/uri/uri.hpp [hunk 3]
```diff
@@ -215,68 +215,125 @@ class uri {
...
    */
   template <typename IntT>
-  optional<IntT> port(typename std::is_integral<IntT>::type * = 0) const {
-    if (auto p = port()) {
-      try {
-        return std::stoi(string_type(std::begin(*p), std::end(*p)));
-      } catch (std::exception &) {
-        return optional<IntT>();
-      }
+  IntT port(typename std::is_integral<IntT>::type * = 0) const {
```
Comment (glynos) [inline src/network/uri/uri.hpp]: Re: your comment the port accessor:

Remember that this already an overload. "string_view port() const" also exists. I provided another accessor that does the conversion to an integer type. This second function might either return an optional, or throw an exception.

Comment (deanberris) [inline src/network/uri/uri.hpp]: Ah, yes. I think throwing here is fine.

**Comment by deanberris** [comment]
LGTM

**Pull request merged**